# Import Library yang dibutuhkan

In [ ]:
print(__doc__)

import warnings
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt
import sklearn
from sklearn.neural_network import MLPClassifier
from sklearn.neural_network import MLPRegressor
import matplotlib.pyplot as plt
import sklearn.metrics as metric
from sklearn.model_selection import train_test_split
from sklearn.neural_network import MLPClassifier
from sklearn.preprocessing import MinMaxScaler
from sklearn import datasets
from sklearn.exceptions import ConvergenceWarning
 




Automatically created module for IPython interactive environment


Disini, kami menggunakan library from sklearn.neural_network import MLPClassifier untuk melakukan klasifikasi menggunakan Neural Network

# Upload Dataset

Dataset Rain in Australia berisi tentang pengamatan cuaca harian pada daerah - daerah di seluruh Australia yang berisi 145460.

Dataset ini memiliki variabel target yaitu RainTomorow yaitu prediksi hujan pada esok hari

In [ ]:
from google.colab import files
uploaded = files.upload() 

Saving weatherAUS.csv to weatherAUS.csv


In [ ]:
df = pd.read_csv("weatherAUS.csv")
df

Date Location  MinTemp  ...  Temp3pm  RainToday  RainTomorrow
0       2008-12-01   Albury     13.4  ...     21.8         No            No
1       2008-12-02   Albury      7.4  ...     24.3         No            No
2       2008-12-03   Albury     12.9  ...     23.2         No            No
3       2008-12-04   Albury      9.2  ...     26.5         No            No
4       2008-12-05   Albury     17.5  ...     29.7         No            No
...            ...      ...      ...  ...      ...        ...           ...
145455  2017-06-21    Uluru      2.8  ...     22.4         No            No
145456  2017-06-22    Uluru      3.6  ...     24.5         No            No
145457  2017-06-23    Uluru      5.4  ...     26.1         No            No
145458  2017-06-24    Uluru      7.8  ...     26.0         No            No
145459  2017-06-25    Uluru     14.9  ...     20.9         No           NaN

[145460 rows x 23 columns]

# Data Preprocessing

Mengubah data kategorikal menjadi data numerikal dengan menggunakan cat codes 

In [ ]:
for col in df.columns:
    df[col] = df[col].astype('category')
    df[col] = df[col].cat.codes

Menghapus data yang null

In [ ]:
for col in df.columns:
    df = df[df[col].notnull()]

In [ ]:
df.head()

Date  Location  MinTemp  MaxTemp  ...  Temp9am  Temp3pm  RainToday  RainTomorrow
0   396         2      213      262  ...      224      262          0             0
1   397         2      153      284  ...      227      287          0             0
2   398         2      208      290  ...      265      276          0             0
3   399         2      171      313  ...      236      309          0             0
4   400         2      254      356  ...      233      341          0             0

[5 rows x 23 columns]

Inisiasi variable target yaitu RainTomorrow.

In [ ]:
df.dropna(inplace=True) # Drop missing values
df.describe(include='all')

X, y = df.drop(columns=['RainTomorrow']), df['RainTomorrow'] # Membagi data antara input (X) dan output (y)
X

Date  Location  MinTemp  MaxTemp  ...  Cloud3pm  Temp9am  Temp3pm  RainToday
0        396         2      213      262  ...        -1      224      262          0
1        397         2      153      284  ...        -1      227      287          0
2        398         2      208      290  ...         2      265      276          0
3        399         2      171      313  ...        -1      236      309          0
4        400         2      254      356  ...         8      233      341          0
...      ...       ...      ...      ...  ...       ...      ...      ...        ...
145455  3431        41      107      267  ...        -1      156      268          0
145456  3432        41      115      286  ...        -1      164      289          0
145457  3433        41      133      302  ...        -1      180      305          0
145458  3434        41      157      303  ...         2      206      304          0
145459  3435        41      228       -1  ...         8      205      253          0

[145460 rows x 22 columns]

In [ ]:
y

0        0
1        0
2        0
3        0
4        0
        ..
49731    0
49732    1
49733    0
49734    0
49735   -1
Name: RainTomorrow, Length: 49736, dtype: int8

DATA TRAINING 90&

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=10)

# UJI COBA NN

Untuk mencari parameter MLP yang terbaik, kami melakukan klasifikasi dedngan berbagai macam variasi. Variasi tersebut antara lain hidden_layer_sizes, learning_rate_init, activation, solver, max_iter. 

Berikut ini salah satu variasi hidden_layer_sizes= (1024,512,256) learning_rate_init=0.001, activation='relu', solver='adam', max_iter=500). 

In [ ]:
from sklearn.neural_network import MLPClassifier

mlp = MLPClassifier(hidden_layer_sizes=(1024,512,256), learning_rate_init=0.001, activation='relu', solver='adam', max_iter=500)
mlp.fit(X_train,y_train)

predict_train = mlp.predict(X_train)
predict_test = mlp.predict(X_test)

In [ ]:
from sklearn.metrics import classification_report,confusion_matrix
print(confusion_matrix(y_train,predict_train))
print(classification_report(y_train,predict_train))

[[  653   354   138]
 [  128 33014   785]
 [  149  4950  4591]]
              precision    recall  f1-score   support

           0       0.70      0.57      0.63      1145
           1       0.86      0.97      0.91     33927
           2       0.83      0.47      0.60      9690

    accuracy                           0.85     44762
   macro avg       0.80      0.67      0.72     44762
weighted avg       0.85      0.85      0.84     44762



In [ ]:
print(confusion_matrix(y_test,predict_test))
print(classification_report(y_test,predict_test))

[[  66   49   18]
 [  16 3568  140]
 [  19  610  488]]
              precision    recall  f1-score   support

           0       0.65      0.50      0.56       133
           1       0.84      0.96      0.90      3724
           2       0.76      0.44      0.55      1117

    accuracy                           0.83      4974
   macro avg       0.75      0.63      0.67      4974
weighted avg       0.82      0.83      0.81      4974



> Berikut ini salah satu variasi (hidden_layer_sizes=(128,64,32), learning_rate_init=0.1 ,activation='logistic', solver='adam', max_iter=500).





In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=10)

In [ ]:
from sklearn.neural_network import MLPClassifier

mlp = MLPClassifier(hidden_layer_sizes=(128,64,32), learning_rate_init=0.1 ,activation='logistic', solver='adam', max_iter=500)
mlp.fit(X_train,y_train)

predict_train = mlp.predict(X_train)
predict_test = mlp.predict(X_test)

In [ ]:
from sklearn.metrics import classification_report,confusion_matrix
print(confusion_matrix(y_train,predict_train))
print(classification_report(y_train,predict_train))

[[    0   884     0]
 [    0 26436     0]
 [    0  7495     0]]
              precision    recall  f1-score   support

           0       0.00      0.00      0.00       884
           1       0.76      1.00      0.86     26436
           2       0.00      0.00      0.00      7495

    accuracy                           0.76     34815
   macro avg       0.25      0.33      0.29     34815
weighted avg       0.58      0.76      0.66     34815



/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [ ]:
print(confusion_matrix(y_test,predict_test))
print(classification_report(y_test,predict_test))

[[    0   394     0]
 [    0 11215     0]
 [    0  3312     0]]
              precision    recall  f1-score   support

           0       0.00      0.00      0.00       394
           1       0.75      1.00      0.86     11215
           2       0.00      0.00      0.00      3312

    accuracy                           0.75     14921
   macro avg       0.25      0.33      0.29     14921
weighted avg       0.56      0.75      0.65     14921



/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


> Berikut ini salah satu variasi (hidden_layer_sizes=(256,128,64), learning_rate_init=0.01 ,activation='logistic', solver='adam', max_iter=500).





In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5, random_state=1)

In [ ]:
from sklearn.neural_network import MLPClassifier

mlp = MLPClassifier(hidden_layer_sizes=(256,128,64), activation='logistic', learning_rate_init=0.01, solver='adam', max_iter=500)
mlp.fit(X_train,y_train)

predict_train = mlp.predict(X_train)
predict_test = mlp.predict(X_test)

In [ ]:
from sklearn.metrics import classification_report,confusion_matrix
print(confusion_matrix(y_train,predict_train))
print(classification_report(y_train,predict_train))

[[   36  1597     0]
 [    0 55015     0]
 [    0 16082     0]]
              precision    recall  f1-score   support

          -1       1.00      0.02      0.04      1633
           0       0.76      1.00      0.86     55015
           1       0.00      0.00      0.00     16082

    accuracy                           0.76     72730
   macro avg       0.59      0.34      0.30     72730
weighted avg       0.59      0.76      0.65     72730



/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [ ]:
from sklearn.metrics import classification_report,confusion_matrix
print(confusion_matrix(y_train,predict_train))
print(classification_report(y_train,predict_train))

[[   36  1597     0]
 [    0 55015     0]
 [    0 16082     0]]
              precision    recall  f1-score   support

          -1       1.00      0.02      0.04      1633
           0       0.76      1.00      0.86     55015
           1       0.00      0.00      0.00     16082

    accuracy                           0.76     72730
   macro avg       0.59      0.34      0.30     72730
weighted avg       0.59      0.76      0.65     72730



/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


> Berikut ini salah satu variasi (hidden_layer_sizes=(256,128,64), activation='logistic', learning_rate_init=0.1, solver='adam', max_iter=500)





In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.7, random_state=1)

In [ ]:
from sklearn.neural_network import MLPClassifier

mlp = MLPClassifier(hidden_layer_sizes=(256,128,64), activation='logistic', learning_rate_init=0.1, solver='adam', max_iter=500)
mlp.fit(X_train,y_train)

predict_train = mlp.predict(X_train)
predict_test = mlp.predict(X_test)

In [ ]:
from sklearn.metrics import classification_report,confusion_matrix
print(confusion_matrix(y_train,predict_train))
print(classification_report(y_train,predict_train))

[[    0   978     0]
 [    0 33056     0]
 [    0  9604     0]]
              precision    recall  f1-score   support

          -1       0.00      0.00      0.00       978
           0       0.76      1.00      0.86     33056
           1       0.00      0.00      0.00      9604

    accuracy                           0.76     43638
   macro avg       0.25      0.33      0.29     43638
weighted avg       0.57      0.76      0.65     43638



/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [ ]:
from sklearn.metrics import classification_report,confusion_matrix
print(confusion_matrix(y_train,predict_train))
print(classification_report(y_train,predict_train))

[[    0   978     0]
 [    0 33056     0]
 [    0  9604     0]]
              precision    recall  f1-score   support

          -1       0.00      0.00      0.00       978
           0       0.76      1.00      0.86     33056
           1       0.00      0.00      0.00      9604

    accuracy                           0.76     43638
   macro avg       0.25      0.33      0.29     43638
weighted avg       0.57      0.76      0.65     43638



/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


# KESIMPULAN

Variasi parameter yang terbaik adalah hidden_layer_sizes= (1024,512,256) learning_rate_init=0.001, activation='relu', solver='adam', max_iter=500) **akurasi yang didapatkan adalah 0.85.**